### ISSUE TITLE
There is no way of detecting if session was reused with synchronous client.

### ISSUE THREAD

**Issue by mfrlin** [eclipse/paho.mqtt.rust#36]
When connecting if you set `clean_session` to `false` the session will be reused if server has the session. So resubscribing is unnecessary. But there is no way of detecting that so you must resubscribe anyway.

Furthermore the example is wrong. You should start consuming even before connecting the client. Because if session is reused and messages are waiting for the client those messages will arrive on connect and will be lost if you don't start consuming before.

I've only tested and explored the synchronous client so I don't know if the same holds for asynchronous client.

**Comment by fpagliughi** [comment]
Yes, sorry. This is a known issue.

**Comment by fpagliughi** [comment]
I put up an initial attempt at a solution. It's in the 'develop' branch at the moment.

There's a new `ServerResponse` enum in token.rs:

```
#[derive(Debug, Clone)]
pub enum ServerResponse {
    /// The granted QoS of the subscription
    Subscribe(i32),
    /// The granted QoS of all the subscriptions
    SubscribeMany(Vec<i32>),
    /// The destination
    Publish(String),
    /// The server URI, MQTT version, and whether the session is present
    Connect(String, i32, bool),
    /// No response from the server
    None,
}
```
This is now the Future `Item` type for Token's, and thus is returned by the Tokens upon successful completion.  The synchronous client can use it like this:

```
match cli.connect(conn_opts) {
    Ok(mqtt::ServerResponse::Connect(server_uri, ver, session_present)) => {
        println!("Connected to: '{}' with MQTT version {}", server_uri, ver);
        if !session_present {
                if let Err(e) = cli.subscribe_many(&subscriptions, &qos) { ... }
```

This is currently working. 

But after testing it out for a few hours, I'm not entirely satisfied with the outcome. After you get the server response, you need to pattern match for the type of response that you expect. For example, after calling `connect()` you need to pattern match to a `ServerResponse::Connect` variety.

What else could it be?

To "fix" this would require having different Token types, or a generic Token type, with each different call returning a different Token. So a `connect()` call would return a `ConnectToken` or `Token<Connect>`.

I will mess with this further to see if one of these feels more natural. But any tests and comments would be appreciated

**Comment by mfrlin** [comment]
I don't know enough Rust yet to provide input. Currently `Ok` variant in `connect` and `reconnect` does not contain any data and I think your current solution provides that data so I'm fine with it.

**Comment by fpagliughi** [comment]
OK. Basically, since there is only one Token type, it returns the server response as an Enum, with the type determined by the request (connect, subscribe, publish, or publish_many). So when you call _connect()_ the token it returns should _always_ give you back the _ServerResponse::Connect_ variant.

Since the app gets back a generic response, it is forced to pattern match to check the result and extract the data from it. That results in a few more lines of code in the app for each call, and they're kind of ugly lines. 

So maybe if I make the library more specific then the applications would be shorter and not so ugly.

**Comment by fpagliughi** [comment]
I started adding support for specific server responses based on the type of request, in a new `srvr_rsp` branch.

For the `AsyncClient` this called for tokens that were specific to the request, like `ConnectToken`, `SubscribeToken`, etc.  Each token's data Item is matched to the response by the server for that request. The `ConnectToken`, for example, returns the URI of the server and whether the session is present on the server.

This still needs to be extended to the async callbacks as well, especially `on_success`, so that the app could have an `on_connect_success()` callback that got the server responses.

For the sync client this just maps out to each call returning a result where the Ok data is the server response, like:
```
pub fn connect<T>(&self, opt_opts:T) -> MqttResult<(String, i32, bool)>
pub fn publish(&self, msg: Message) -> MqttResult<String>
pub fn subscribe(&self, topic: &str, qos: i32) -> MqttResult<i32>
``` 
For connecting you would get back a tuple: `(server_uri, mqtt_ver, session_present)`

I'm still a little unsure whether this is the proper way to proceed. There's a lot more library code to support a small API improvement. I'm still trying to figure out a way to replace a lot of the new, redundant code with something generic.

**Comment by fpagliughi** [comment]
Fixed in 5aa49de8e8dd6d5ff45296b6f79056e488ab18aa with additional modifications after that in the 'develop' branch.
